marlowe-contracts/test/Spec/Marlowe/Contracts.hs

```diff
@@ -0,0 +1,395 @@
+  , testCase "Future Contract (with margins calls)" futureWithMarinCall
+  , testCase "American Call Option test" americanCallOptionTest
+  , testCase "American Call Option (exercised) test" americanCallOptionExercisedTest
+  , testCase "European Call Option test" europeanCallOptionTest
+  , testCase "European Call Option (exercised) test" europeanCallOptionExercisedTest
+  , testCase "Reverse Convertible test" reverseConvertibleTest
+  , testCase "Reverse Convertible (exercised) test" reverseConvertibleExercisedTest
+  ]
+
+w1Pk, w2Pk :: Party
+w1Pk = Role "party1"
+w2Pk = Role "party2"
+
+tokName :: TokenName
+tokName = "testcoin"
+
+tokSymbol :: CurrencySymbol
+tokSymbol = ""
+
+tok :: Token
+tok = Token tokSymbol tokName
+
+tokValueOf :: Integer -> Money
+tokValueOf = singleton tokSymbol tokName
+
+assertTotalPayments :: Party -> [Payment] -> Money -> Assertion
+assertTotalPayments p t x = assertBool "total payments to party" (totalPayments t == x)
+  where
+    totalPayments = mconcat . map (\(Payment _ _ a) -> a) . filter (\(Payment _ a _) -> a == Party p)
+
+assertNoWarnings :: [a] -> Assertion
+assertNoWarnings [] = pure ()
+assertNoWarnings t  = assertBool "No warnings" $ null t
+
+assertClose :: Contract -> Assertion
+assertClose = assertBool "Contract is in Close" . (Close==)
+
+step :: TransactionInput -> ([TransactionWarning], [Payment], State, Contract) -> Either TransactionError ([TransactionWarning], [Payment], State, Contract)
+step i (_,_,s,c) =
+  case computeTransaction i s c of
+    Error err                   -> Left err
+    TransactionOutput w p s' c' -> Right (w, p, s', c')
+
+-- |Zero-coupon Bond test
+zeroCouponBondTest :: IO ()
+zeroCouponBondTest =
+  let zcb =
+        zeroCouponBond
+          w1Pk
+          w2Pk
+          (Slot 100)
+          (Slot 200)
+          (Constant 75_000_000)
+          (Constant 90_000_000)
+          ada
+          Close
+   in either left right $
+        step (TransactionInput (0, 0)     [NormalInput $ IDeposit w1Pk w1Pk ada 75_000_000]) ([], [], emptyState 0, zcb)
+    >>= step (TransactionInput (100, 110) [NormalInput $ IDeposit w1Pk w2Pk ada 90_000_000])
+  where
+    left err = assertFailure $ "Transactions are not expected to fail: " ++ show err
+    right (w,p,_,c) = do
+      assertClose c
+      assertNoWarnings w
+      assertTotalPayments w1Pk p (lovelaceValueOf 90_000_000)
+
+-- |Swap contract test
+swapContractTest :: IO ()
+swapContractTest =
+  let backSwap =
+        swap w1Pk ada (Constant 10_000_000) w2Pk tok (Constant 30) (Slot 100) $
+        swap w2Pk ada (Constant 10_000_000) w1Pk tok (Constant 30) (Slot 200) Close
+   in either left right $
+        step (TransactionInput (0, 0)
+              [ NormalInput $ IDeposit w1Pk w1Pk ada 10_000_000
+              , NormalInput $ IDeposit w2Pk w2Pk tok 30
+              , NormalInput $ IDeposit w2Pk w2Pk ada 10_000_000
+              , NormalInput $ IDeposit w1Pk w1Pk tok 30
```

Comment:
Yes, that way we could do this in separate transactions too, for example, if we wanted